README.md
The image processing pipelines that the HiRISE team operates
internally to produce higher order products are more than just the
'simple' programs available in ISIS3.  Those processes that run in
the HiRISE Operations Center (HiROC) are a complicated dance of
primarily Perl and ISIS3 run by a custom job management system, all
of which is based on the HiRISE catalog (HiCat) database.

This makes the HiROC system excellent for processing the Gigabytes of new
data that arrive daily from Mars, and allow the team to perform massive
reprocessing of the entire data set, as needed, and to produce on a large
scale a variety of derived data products.

However, that same complexity makes it difficult to reproduce exactly
what that system is doing on a small scale.

The programs in this repo are meant to replicate the HiRISE processing
chain on a small scale, so that individual algorithms and processes can
be investigated, without needed a massive data processing system.

The programs here have similar names to HiRISE pipelines, but only focus
on the data processing.  The HiRISE pipeline programs do a lot of other 
tasks relevant to being part of a massive ground data system, and clearly,
those functionalities aren't replicated here.

The HiROC system begins by watching the MRO project's raw data server for
new products with the FEI_Watchdog program, and then the HiDog pipeline 
fetches those products down to HiROC and the the EDRgen pipeline converts 
them into .img EDR products.

Since that is the most basic form of the data available from the PDS, we 
will start there, and assume that you have downloaded a set of EDR .img
files from the PDS.

We will use simple sqlite files instead of a full-up relational database.

# External Dependencies

These programs use as much 'vanilla' Python 3 as possible.

However, the use of the pvl library (https://pvl.readthedocs.io).


# EDR_Stats

* Convert a HiRISE EDR Product to an ISIS cube for subsequent pipeline processing using the ISIS system for much of the work.
* Gather image statistics about the observation and update HiCat's EDR_Products table with those statistics
* Create an image histogram and store it as an ASCII file for use by other applications.
* Calculate the Signal-to-Noise Ratio (SNR).

# HiCal

* Perform a radiometric calibration correction and conversion to "I/F" units. The calibration corrects for the variable gain and dark current of each CCD detector column.
* As part of the radiometric calibration, the current pipeline implementation applies a separate gain line-drift correction. This original line-drift correction is used in lieu of the correction found in the current ISIS "hical" program as the original correction algorithm has been demonstrated to work better.
* Perform furrow correction. The image columns at the channel join are checked for furrows. Pixels in the furrow region (at the channel join) whose DN values have gone above a threshold are set to the null pixel value as these pixels can not be calibrated. If a RED-filter image experiences furrowing, a rare event usually caused by an improperly commanded observation, then the furrowed pixels will be permanently set to null pixels for all of the HiRISE standard and extras products. For BG and IR-filter images that make up the color products, the furrowed pixels will be interpolated in the HiColorNorm pipeline step.
* If an observation is determined to have furrows then an entry is made in HiCat's Tags table to indicate the level of furrowing that has occurred. The comment field in the Tag table entry contains a number indicating the percent of the first furrow column that had pixel values above the threshold furrow value.
* Due to HiRISE instrument instability problems, the HiCal pipeline performs a noise reduction procedure to reduce the number of bad pixels in an image observation. The noise correction is applied when the standard deviation of the dark pixel or mask regions exceed a threshold, or if the number of LIS pixels exceeds a threshold.
* A high-pass filter "cubenorm" step is applied to the calibrated image. Due to camera instabilities, residual vertical stripping often exists in the imaging that is corrected by this empirical method. The average standard deviation of this change is calculated and stored in HiCat.
* The ISIS program "hidestripe" is applied to suppress horizontal stripping seen whenever an observation is acquired using mixed-binning commanding. The standard deviation of this change is calculated and started in HiCat.
* Browse and thumbnail images of the EDR channel products are created.
* Entries for the browse and thumbnail EDR products are added to HiCat's Extras_Products table.
* An entry is made to the HiStitch pipeline sources table when two channel products of the same CCD have successfully gone through the HiCal pipeline step.


HISTORY.rst
=========
 History
=========

All notable changes to this project will be documented in this file.

The format is based on `Keep a Changelog <https://keepachangelog.com/en/1.0.0/>`_,
and this project adheres to `Semantic Versioning <https://semver.org/spec/v2.0.0.html>`_.

When updating this file, please add an entry for your change under
Unreleased_ and one of the following headings:

- Added - for new features.
- Changed - for changes in existing functionality.
- Deprecated - for soon-to-be removed features.
- Removed - for now removed features.
- Fixed - for any bug fixes.
- Security - in case of vulnerabilities.

If the heading does not yet exist under Unreleased_, then add it
as a 3rd level heading, underlined with pluses (see examples below).

When preparing for a public release add a new 2nd level heading,
underlined with dashes under Unreleased_ with the version number
and the release date, in year-month-day format (see examples below).


Unreleased
----------

Added
+++++
* lisfix: Added the lisfix module.
* bitflips: Sometimes the very end of the histogram (although not a formal minima) is the appropriate choice, and
  is now considered.
* bitflips: Added capability to ignore minor maxima at the ends of the histogram to "roll down" to a better solution.

Changed
+++++++
* name change of project from pyrise to hiproc.
* bitflips: Changed the default medstd_limit from 300 to 400 DN.

Fixed
+++++
* bitflips: There were a variety of edge cases that resulted in errors.  The appropriate guardrails, handlers,
  and recovery logic has now been added.
* HiStitch: The equalize and balance parameters cannot both be true.

0.4.0 (2020-09-22)
------------------
* Tremendous amount of re-working in bitflips to improve
  performance.
* Format cleanup

0.3.0 (2020-05-16)
------------------
* Confirmed that EDR_Stats and HiCal produce identical output cubes.

0.2.0 (2020-05-06)
------------------
* Updated with bit-flip correction.

0.1.0 2020-03-21
----------------
* First shared on GitHub


HiStitch.py
#!/usr/bin/env python
"""Stitch together two HiRISE channel files to create a single CCD image file."""

# Copyright 2019, Ross A. Beyer (rbeyer@seti.org)
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


# This program is based on HiStitch version 1.32 2016/08/04
# and on the Perl HiStitch program: ($Revision: 1.24 $ $Date: 2016/08/05 18:05:28 $)
# by Eric Eliason
# which is Copyright(C) 2016 Arizona Board of Regents, under the GNU GPL.
#
# Since that suite of software is under the GPL, none of it can be directly
# incorporated in this program, since I wish to distribute this software
# under the Apache 2 license.  Elements of this software (written in an entirely
# different language) are based on that software but rewritten from scratch to
# emulate functionality.

import argparse
import re
import shutil
import sys
from pathlib import Path

import hirise
import kalasiris as isis


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--db',          required=False, default='HiCat.db')
    parser.add_argument('-o', '--output', required=False, default='.HiStitch.cub')
    parser.add_argument('-c', '--conf',    required=False, default='HiStitch.conf')
    parser.add_argument('-k', '--keep',   required=False, default=False)
    parser.add_argument('cube0', metavar=".cub-file")
    parser.add_argument('cube1', metavar=".cub-file", required=False)

    # The original Perl needed the specific output of cubenorm from a
    # particular step in the HiCal pipeline before here.  However, rather
    # than keep track of those files, open them, and read them again, we
    # can (and did) perform the relevant check in HiCal.py, and then save
    # that in the db that we can check now.

    args = parser.parse_args()

    ofile_cub = set_outcube(args)

    # GetConfigurationParameters()
    conf = pvl.load(args.conf)

    # GetProductFiles()
    channel_cub = {'0': None, '1': None}
    if not args.cube1:
        channel_cub['0'] = Path(args.cube0)
    else:
        channel_cub = sort_input_cubes(args.cube0, args.cube1)

    # PrepareDBStatements()
    #   select from HiCat.EDR_Products
    db0
    db1

    (truthchannel, balanceratio) = HiStitch(channel_cub['0'], channel_cub['1'],
                                            ofile_cub,
                                            conf['HiStitch'], db0, db1,
                                            ccdchan,
                                            keep=args.keep)

    # insert ObsID, ccdchan[0], truthchannel, balanceratio
    # into HiCat.CCD_Processing_Statistics


def set_outcube(args) -> Path:
    if args.output.startswith('.'):
        try:
            oid_a = hirise.ObservationID(args.cube0)
            if args.cube1:
                oid_b = hirise.ObservationID(args.cube1)
                if oid_a != oid_b:
                    raise ValueError
            return Path(args.cube0).with_name(oid_a + args.output)
        except ValueError:
            # Couldn't get ObsIDs from filenames, look in the headers?
            pid_a = isis.getkey_k(args.cube0, 'Archive', 'ProductId')
            if args.cube1:
                pid_b = isis.getkey_k(args.cube1, 'Archive', 'ProductId')
                if pid_a != pid_b:
                    raise ValueError(f'These do not appear to be channels '
                                     'from the same Observation: {args.cub1}: '
                                     '{pid1} and {args.cub2}: {pid2}')
            return Path(args.cube0).with_name(pid_a + args.output)
    else:
        return Path(args.output)


def get_id_from_cube(


def sort_input_cubes(a_cub: os.PathLike, b_cub: os.PathLike) -> dict:
    '''Figures out which one is Channel 0 and which is Channel 1.'''

    d = {'0': None, '1': None}

    a_chan = getkey_k(a_cub, 'Instrument', 'ChannelNumber')
    b_chan = getkey_k(b_cub, 'Instrument', 'ChannelNumber')

    try:
        d[a_chan] = Path(a_cub)
        d[b_chan] = Path(b_cub)
        for k, v in d.items():
            if v is None:
                raise RuntimeError(f'These have the same channel: {cube1_chan}.')
    except KeyError:
        raise RuntimeError(f'This file had a channel other than 0 or 1: {err}')

    return d


def HiStitch(cub0, cub1, out_cub, conf, db0, db1, ccdchan, keep=False) -> tuple:
    ccd_number = int(re.match(r"RED|IR|BG(\d+)", ccdchan[0]).group())

    # Allows for indexing in lists ordered by bin value.
    b = 1, 2, 4, 8, 16

    # This string will get placed in the filename for all of our
    # temporary files. It will (hopefully) prevent collisions with
    # existing files and also allow for easy clean-up if keep=True
    temp_token = datetime.now().strftime('HiStitch-%y%m%d%H%M%S')

    # ProcessingStep() - mostly sets up stuff
    max_mean = max(db0.IMAGE_MEAN, db1.IMAGE_MEAN)
    flags = set_flags(conf, db0, db1, ccd_number, b.index(db0.bin), max_mean)

    # HiStitchStep() - runs HiStitch, and inserts to db
    HiStitchStep(cub0, cub1, out_cub, db0.bin, ccd_number,
                 flags.balance, flags.equalize)

    if cub1 and balance:
        # run getkey from ?HiStitch_output?
        truthchannel = isis.getkey_k(stitch_cub, 'HiStitch', 'TruthChannel')
        balanceratio = isis.getkey_k(stitch_cub, 'HiStitch', 'BalanceRatio')

    if flags.furrow:
        furrow_file = out_cube.with_suffix(f'.{temp_token}.temp.cub')
        HiFurrow_Fix(stitch_cub, furrow_file, max_mean, keep=keep)
        furrow_file.rename(out_cube)

    if cub1 and balance:
        return(truthchannel, balanceratio)
    else:
        return(None, None)


def set_flags(conf, db0, db1, ccdnum: int, bindex: int, max_mean) -> collections.namedtuple:
    '''Set various processing flags based on various configuration
    parameters.'''
    HiStitchFlags = collections.namedtuple('HiStitchFlags',
                                           ['furrow', 'balance', 'equalize'])

    max_lis = max(db0.LOW_SATURATED_PIXELS, db1.LOW_SATURATED_PIXELS)
    max_std = max(db0.CAL_MASK_STANDARD_DEVIATION,
                  db1.CAL_MASK_STANDARD_DEVIATION)
    max_dstd = max(db0.IMAGE_DARK_STANDARD_DEVIATION,
                   db1.IMAGE_DARK_STANDARD_DEVIATION)
    max_gper = max(db0.GAP_PIXELS_PERCENT, db1.GAP_PIXELS_PERCENT)

    # We don't run FurrowCheck() here, but do it earlier in HiCal.py
    # and store the result to get now.

    # Determine if the Furrow Normalization is to occur
    furrow = False
    if(db0.zap or db1.zap
       and max_mean >= conf['HiStitch_Furrow_Normalization']
       and (db0.bin == 2 or db0.bin == 4)):
        furrow = True

    # Determine if the balance or equalize options are to occur
    balance = False
    equalize = False

    if (conf['HiStitch_Balance'] or conf['HiStitch_Equalize'] or
        (conf['HiStitch_Balance_Processing'][ccdnum] == 0 or
         (max_lis < conf['HiStitch_LIS_Pixels'] and
          # How its written below is how the original had it, but I think
          # it needs to be indexed by bindex:
          # max_std < conf['HiStitch_Balance_Bin_Mask_STD'][bindex] and
          # max_dstd < conf['HiStitch_Balance_Bin_DarkPixel_STD'][bindex]))):
          max_std < conf['HiStitch_Balance_Bin_Mask_STD'][db0.bin] and
          max_dstd < conf['HiStitch_Balance_Bin_DarkPixel_STD'][db0.bin]))):
        equalize = conf['HiStitch_Equalize'] and max_gper < conf['HiStitch_Gap_Percent']
        balance = conf['HiStitch_Balance']

    return HiStitchFlags(furrow, balance, equalize)


def HiStitchStep(in_cub0, in_cub1, out_cub, binning, ccdnum, balance, equalize):
    zpad_bin = '{:0>2}'.format(binning)
    histitch_args = {'from1': in_cub0, 'to': out_cub}
    if in_cub1:
        histitch_args['from2'] = in_cub1
    if not balance and not equalize:
        histitch_args['balance'] = False
    if equalize or balance:
        histitch_args['skip'] = conf[f'HiStitch_Bin{zpad_bin}_Area'][0]
        histitch_args['seamsize'] = conf[f'HiStitch_Bin{zpad_bin}_Area'][1]
        histitch_args['channel'] = conf['HiStitch_Control_Channel'][ccdnum]
        if balance:
            histitch_args['balance'] = 'TRUE'
        if equalize:
            histitch_args['balance'] = 'EQUALIZE'
            histitch_args['width'] = ''
            histitch_args['operator'] = ''

    return isis.histitch(**histitch_args)


def HiFurrow_Fix(in_cub, out_cub, max_mean, keep=False):
    '''Perform a normalization of the furrow region of bin 2 or 4
       HiRISE images. The input to this script is a HiRISE stitch
       product containing both channels of a CCD.'''

    binning = isis.getkey_k(in_cub, 'Instrument', 'Summing')
    lines = isis.getkey_k(in_cub, 'Dimensions', 'Lines')
    samps = isis.getkey_k(in_cub, 'Dimensions', 'Samples')

    if binning != '2' and binning != '4':
        raise Exception('HiFurrow_Fix only supports correction for '
                        'bin 2 or 4 data.')
    if binning == '2' and samps != '1024':
        raise Exception('Improper number of samples, {} for a stitch '
                        'product with bin 2.'.format(samps))

    # This string will get placed in the filename for all of our
    # temporary files. It will (hopefully) prevent collisions with
    # existing files and also allow for easy clean-up if keep=True
    temp_token = datetime.now().strftime('HFF-%y%m%d%H%M%S')
    to_del = PathSet()

    # For bin2 and bin4 imaging, specify width of furrow based on
    # image average DN range
    range_low['2'] = (512, 513)  # 2 pixel furrow width
    range_mid['2'] = (511, 514)  # 4 pixel furrow width
    range_hgh['2'] = (511, 514)  # 4 pixel furrow width
    range_max['2'] = (510, 515)  # 6 pixel furrow width

    # Original code had low/mid/hgh for bin2 and bin4, but they
    # were hard-coded to be identical.
    dn_range_low = 9000
    dn_range_mid = 10000
    dn_range_hgh = 12000

    range_low['4'] = (256, 257)  # 2 pixel furrow width
    range_mid['4'] = (255, 258)  # 4 pixel furrow width
    range_hgh['4'] = (255, 258)  # 4 pixel furrow width
    range_max['4'] = (254, 259)  # 6 pixel furrow width

    if max_avg > dn_range_hgh:
        dn_range = range_max[binning]
    elif max_avg > dn_range_mid:
        dn_range = range_hgh[binning]
    elif max_avg > dn_range_low:
        dn_range = range_mid[binning]
    else:
        dn_range = range_low[binning]

    lpf_samp = int((dn_range[1] - dn_range[0] + 1) / 2) * 4 + 1
    lpf_line = int(lpf_samp / 2) * 20 + 1

    # Create a mask file
    # DN=1 for non-furrow area
    # DN=0 for furrow area
    fx_cub = to_del.add(in_cub.with_suffix(f'.{temp_token}.fx.cub'))
    isis.fx(to=fx_cub, mode='OUTPUTONLY', lines=lines, samples=samps,
            equation=f'(1*(sample<{dnrange[0]})+ 1*(sample>{dn_range[1]}) + 0)')

    # Create a file where the furrow area is set to null
    mask1_cub = to_del.add(in_cub.with_suffix(f'.{temp_token}.mask1.cub'))
    isis.mask(in_cub, mask=fx_cub, to=mask1_cub, min_=1, max_=1,
              preserve='INSIDE', spixels='NULL')

    # Lowpass filter to fill in the null pixel area
    lpf_cub = to_del.add(in_cub.with_suffix(f'.{temp_token}.lpf.cub'))
    isis.lowpass(mask1_cub, to=lpf_cub, sample=lpf_samp, line=lpf_line,
                 null=True, hrs=False, his=False, lis=False)

    # Create a file where non-furrow columns are set to null
    mask2_cub = to_del.add(in_cub.with_suffix(f'.{temp_token}.mask2.cub'))
    isis.mask(in_cub, mask=fx_cub, to=mask2_cub, min_=0, max_=0,
              preserve='INSIDE', spixels='NULL')

    # Highpass filter the furrow region
    hpf_cub = to_del.add(in_cub.with_suffix(f'.{temp_token}.hpf.cub'))
    isis.highpass(mask2_cub, to=hpf_cub, sample=1, line=lpf_line)

    # Add lowpass and highpass together to achieve desired result
    alg_cub = to_del.add(in_cub.with_suffix(f'.{temp_token}.alg.cub'))
    isis.algebra(from_=lpf_cub, from2=hpf_cub, to=alg_cub,
                 operator='ADD', A=1.0, B=1.0)

    # copy the input file to the output file then mosaic the
    # furrow area as needed.
    shutil.copyfile(in_cub, out_cube)
    isis.handmos(alg_cub, mosaic=out_cube, outsample=1, outline=1, outband=1,
                 insample=1, inline=1, inband=1, create='NO')

    if not keep:
        to_del.unlink()

    return


if __name__ == "__main__":
    main()
